hopp/tools/design/wind/power_curve_tools.py

```diff
@@ -0,0 +1,174 @@
+        raise UserWarning("The length of the wind speed and power vectors must be the same")
+    rotor_area = np.pi*((rotor_diameter/2)**2)
+    if isinstance(wind_speeds_ms, list):
+        wind_speeds_ms = np.array(wind_speeds_ms)
+    if isinstance(power_curve_kw, list):
+        power_curve_kw = np.array(power_curve_kw)
+    
+    # power available in the wind (kW)
+    p_wind = 0.5*air_density*rotor_area*(wind_speeds_ms**3)/1e3
+    cp = power_curve_kw/p_wind
+    cp = np.where(cp < 0, 0, cp)
+    return list(cp)
+
+def calculate_power_from_cp(wind_speeds_ms, cp_curve, rotor_diameter, rated_power_kW, air_density = 1.225):
+    """Calculate power curve from power coefficient curve (Cp).
+
+    Args:
+        wind_speeds_ms (List | np.ndarray): list of wind speeds in m/s
+        cp_curve (List | np.ndarray): power curve coefficients (Cp) at each wind speed in `wind_speeds_ms`
+        rotor_diameter (float): rotor diameter of the turbine in meters.
+        air_density (float, Optional): Air density assumed for power-curve calculations in kg/m3.
+             Defaults to 1.225.
+    
+    Raises:
+        UserWarning: if `wind_speeds_ms` and `cp_curve` are different lengths.
+
+    Returns:
+        list: turbine power (in kW) at each wind speed in `wind_speeds_ms`
+    """
+
+    if len(wind_speeds_ms) != len(cp_curve):
+        raise UserWarning("The length of the wind speed and coefficient of power vectors must be the same")
+
+    rotor_area = np.pi*((rotor_diameter/2)**2)
+    if isinstance(wind_speeds_ms, list):
+        wind_speeds_ms = np.array(wind_speeds_ms)
+    if isinstance(cp_curve, list):
+        cp_curve = np.array(cp_curve)
+    
+    # power available in the wind (kW)
+    p_wind = 0.5*air_density*rotor_area*(wind_speeds_ms**3)/1e3
+    power_kW = cp_curve*p_wind
+    power_kW = np.where(power_kW > rated_power_kW, rated_power_kW, power_kW)
+    power_kW = np.where(power_kW < 0, 0, power_kW)
+    return list(power_kW)
+
+def estimate_thrust_coefficient(wind_speeds_ms, cp_curve, plot=False, print_output=False):
+    """Calculate thrust coefficient curve (Ct) from power coefficient curve (Cp).
+
+    Args:
+        wind_speeds_ms (List | np.ndarray): list of wind speeds in m/s
+        cp_curve (List | np.ndarray): power curve coefficients (cp) at each wind speed in `wind_speeds_ms`
+        plot (bool, Pptional): whether to plot Cp and Ct curve. Defaults to False.
+        print_output (bool, Optional): Whether to print Cp and Ct curves. Defaults to False.
+    
+    Raises:
+        UserWarning: if `wind_speeds_ms` and `cp_curve` are different lengths.
+
+    Returns:
+        list: thrust curve coefficients (Ct) at each wind speed in `wind_speeds_ms`
+    """
+    
+    # Check that the wind speed and the coefficient of power are the same length
+    if len(wind_speeds_ms) != len(cp_curve):
+        raise UserWarning("The length of the wind speed and coefficient of power vectors must be the same")
```

Comment:
I know the `UserWarning` has been used a lot, but I'm just realizing that it may be being misused. I think it's often being used in place of an actual error quite often, so programs will fail silently where they shouldn't be (such as this specific case). In many cases, I think `ValueError` would be more appropriate unless a non-failing warning is what is intended.

Reply:
This instance was changed, but it seems there are quite a few them hanging around. Could update each to a more appropriate warning or error to ensure the code fails when it's necessary rather than simply warning a user? I'm adding my recommendation to each. Feel free to resolve them as they're dealt with or push back where you think it's appropriate.